lib/appoptics_apm/inst/graphql.rb

```diff
@@ -0,0 +1,188 @@
+          send(:remove_const, :AppOpticsTracing)
+        else
+          dont_redefine = true
+        end
+      end
+
+      unless dont_redefine
+        #-----------------------------------------------------------------------------#
+        #----- this class is duplicated in the graphql gem ---------------------------#
+        #-----------------------------------------------------------------------------#
+        class AppOpticsTracing < GraphQL::Tracing::PlatformTracing
+          # These GraphQL events will show up as 'graphql.prep' spans
+          PREP_KEYS = ['lex', 'parse', 'validate', 'analyze_query', 'analyze_multiplex'].freeze
+
+          # During auto-instrumentation this version of AppOpticsTracing is compared
+          # with the version provided in the graphql gem, so that the newer
+          # version of the class can be used
+          VERSION = Gem::Version.new('1.0.0').freeze
+
+          self.platform_keys = {
+            'lex' => 'lex',
+            'parse' => 'parse',
+            'validate' => 'validate',
+            'analyze_query' => 'analyze_query',
+            'analyze_multiplex' => 'analyze_multiplex',
+            'execute_multiplex' => 'execute_multiplex',
+            'execute_query' => 'execute_query',
+            'execute_query_lazy' => 'execute_query_lazy'
+          }
+
+          def platform_trace(platform_key, _key, data)
+            return yield if !defined?(AppOpticsAPM) || gql_config[:enabled] == false
+
+            kvs = metadata(data)
+            kvs[:Key] = platform_key if PREP_KEYS.include?(platform_key)
+
+            maybe_set_transaction_name(kvs[:InboundQuery]) if kvs[:InboundQuery]
+
+            ::AppOpticsAPM::SDK.trace(span_name(platform_key), kvs) do
+              kvs.clear # we don't have to send them twice
+              result = yield
+              report_errors(result)
+              result
+            end
+          end
+
+          def platform_field_key(type, field)
+            "graphql.#{type.name}.#{field.name}"
+          end
+
+          private
+
+          def gql_config
+            ::AppOpticsAPM::Config[:graphql] ||= {}
+          end
+
+          ###
+          # any errors graphql has dealt with are added to the response
+          def report_errors(result)
+            return unless result.is_a?(Array)
+
+            result.each do |res|
+              if res.is_a?(GraphQL::Query::Result) && res.to_h['errors']
+                msg = res.to_h['errors'].map { |r| r['message'] }.join("\n")
+                AppOpticsAPM::SDK.log_info(Message: "GraphQL Errors:\n#{msg}")
+              end
+            end
+          end
+
+          def maybe_set_transaction_name(query)
+            return if gql_config[:transaction_name] == false ||
+                      ::AppOpticsAPM::SDK.get_transaction_name
+
+            split_query = query.strip.split(/\W+/, 3)
+            name = "graphql.#{split_query[0..1].join('.')}"
```

Comment:
Would the values in the range [0..1] contain anything that would be removed by the query sanitize (in other words this wouldn't set part of the transaction name to something that query sanitize would be removing for the query field)?

Reply:
It wouldn't, the first 2 words can't contain any parameters, they are basically 'query', 'mutation', or 'subscription' plus a type.

Reply:
okay, thanks for confirming.